### PR TITLE
AdvancedCronJob webhook impl and test cases

### PR DESCRIPTION
**Pull request by rishi-anand**
### Ⅰ. Describe what this PR does
This PR implements webhook and test case for AdvancedCronJob. 
Proposal: https://github.com/openkruise/kruise/blob/master/docs/proposals/20201024-advancedcronjob.md

### Ⅱ. Does this pull request fix one issue?

#215 

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
Create an instance of `AdvancedCronJob` CRD and reconcile it.

### Ⅳ. Describe how to verify it
Can test by create below sample instance of `AdvancedCronJob` CRD
```
apiVersion: apps.kruise.io/v1alpha1
kind: AdvancedCronJob
metadata:
  name: advanced-sample-demo
spec:
  schedule: "* * * * *"
  concurrencyPolicy: Replace
  paused: false
  successfulJobsHistoryLimit: 3
  failedJobsHistoryLimit: 3
  template:
    broadcastJobTemplate:
      spec:
        template:
          spec:
            containers:
              - name: hello
                image: busybox
                args:
                  - /bin/sh
                  - -c
                  - date; echo "Hello from the BroadcastCronJob - SpectroCloud  "
            restartPolicy: Never
        completionPolicy:
          type: Always
          activeDeadlineSeconds: 200
```

### Ⅴ. Special notes for reviews
@Fei-Guo, Implemented webhook and added a test case.

### CODE REVIEW

File: pkg/controller/advancedcronjob/advancedcronjob_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2019 The Kruise Authors.
```
Comment (Fei-Guo) [inline pkg/controller/advancedcronjob/advancedcronjob_controller_test.go:2]: 2019->2020

File: pkg/webhook/advancedcronjob/mutating/advancedcronjob_create_update_handler.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 The Kruise Authors.
```
Comment (Fei-Guo) [inline pkg/webhook/advancedcronjob/mutating/advancedcronjob_create_update_handler.go:2]: 2019->2020

File: pkg/webhook/advancedcronjob/mutating/webhooks.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2019 The Kruise Authors.
```
Comment (Fei-Guo) [inline pkg/webhook/advancedcronjob/mutating/webhooks.go:2]: 2019->2020

File: pkg/webhook/advancedcronjob/validating/advancedcronjob_create_update_handler.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2019 The Kruise Authors.
```
Comment (Fei-Guo) [inline pkg/webhook/advancedcronjob/validating/advancedcronjob_create_update_handler.go:2]: 2019>2020

File: pkg/webhook/advancedcronjob/validating/webhooks.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2019 The Kruise Authors.
```
Comment (Fei-Guo) [inline pkg/webhook/advancedcronjob/validating/webhooks.go:2]: 2019->2020

**Review by Fei-Guo — COMMENTED** [review]
I don't think you need to add req.NamespacedName in all logs. If the reconciler returns an error, the controller runtime should print out the failed request.

File: apis/apps/v1alpha1/advancedcronjob_types.go [hunk 6]
```diff
@@ -38,8 +38,7 @@ type AdvancedCronJobSpec struct {
 	// - "Allow" (default): allows CronJobs to run concurrently;
 	// - "Forbid": forbids concurrent runs, skipping next run if previous run hasn't finished yet;
 	// - "Replace": cancels currently running job and replaces it with a new one
-	// +optional
-	ConcurrencyPolicy ConcurrencyPolicy `json:"concurrencyPolicy,omitempty" protobuf:"bytes,3,opt,name=concurrencyPolicy"`
+	ConcurrencyPolicy ConcurrencyPolicy `json:"concurrencyPolicy" protobuf:"bytes,3,opt,name=concurrencyPolicy"`
```
[the remark this answers is not in the document]
Comment (rishi-anand) [reply apis/apps/v1alpha1/advancedcronjob_types.go:41]: This is a mandatory field. Policy should be either "Allow", "Forbid", "Replace". Old job deletion depends on policy selected by user and user should provide the expected behaviour.

File: apis/apps/v1alpha1/defaults.go [hunk 7]
```diff
@@ -92,6 +92,35 @@ func setSidecarDefaultContainer(sidecarContainer *SidecarContainer) {
...
+
+	if obj.Spec.FailedJobsHistoryLimit == nil {
+		var failedJobHistorylimit int32
+		failedJobHistorylimit = 0
+		obj.Spec.FailedJobsHistoryLimit = &failedJobHistorylimit
+	}
+
+	if obj.Spec.StartingDeadlineSeconds == nil {
+		var deadlineSeconds int64
+		deadlineSeconds = 3600
```
[the remark this answers is not in the document]
Comment (rishi-anand) [reply apis/apps/v1alpha1/defaults.go:119]: If any schedule is missed for given seconds in `startingDeadlineSeconds` then controller will not schedule the job.  I think **I should not set any defaults and if nothing is given then job will be scheduled even if job is not scheduled for longer duration**.

File: config/rbac/role.yaml [hunk 8]
```diff
@@ -211,7 +211,7 @@ rules:
 - apiGroups:
   - batch
   resources:
-  - advancedcronjobs
+  - jobs
```
[the remark this answers is not in the document]
Comment (rishi-anand) [reply config/rbac/role.yaml:214]: That was a mistake in job controller.

_Earlier:_
```
// +kubebuilder:rbac:groups=batch,resources=advancedcronjobs,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=batch,resources=advancedcronjobs/status,verbs=get;update;patch

func (r *ReconcileAdvancedCronJob) reconcileJob(ctx context.Context, req ctrl.Request, advancedCronJob appsv1alpha1.AdvancedCronJob) (ctrl.Result, error) {
	advancedCronJob.Status.Type = appsv1alpha1.JobTemplate
	var childJobs batchv1.JobList
```

_Latest:_
```
// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=batch,resources=jobs/status,verbs=get;update;patch

func (r *ReconcileAdvancedCronJob) reconcileJob(ctx context.Context, req ctrl.Request, advancedCronJob appsv1alpha1.AdvancedCronJob) (ctrl.Result, error) {
	advancedCronJob.Status.Type = appsv1alpha1.JobTemplate
	var childJobs batchv1.JobList
```

Advanced cron job role is already added in top level advancedcronjobs controller.

File: apis/apps/v1alpha1/advancedcronjob_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rishi-anand) [reply apis/apps/v1alpha1/advancedcronjob_types.go]: @Fei-Guo, made it to optional field in commit https://github.com/openkruise/kruise/pull/468/commits/aa5bc2c7c7d02e38c348cbbb7134aa9d8b76c80c as it is an optional field in k8s cron job.

File: apis/apps/v1alpha1/defaults.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (rishi-anand) [reply apis/apps/v1alpha1/defaults.go]: `StartingDeadlineSeconds` is nil as there is no defaults if not provided by user. If user doesn't provide `StartingDeadlineSeconds` then job will be scheduled again after any number of job schedule failures.

File: config/webhook/manifests.yaml [hunk 9]
```diff
@@ -140,6 +158,24 @@ metadata:
...
   name: validating-webhook-configuration
 webhooks:
+- clientConfig:
+    caBundle: Cg==
+    service:
+      name: webhook-service
+      namespace: system
+      path: /validate-apps-kruise-io-v1alpha1-advancedcronjob
+  failurePolicy: Fail
+  name: madvancedcronjob.kb.io
```
[the remark this answers is not in the document]
Comment (rishi-anand) [reply config/webhook/manifests.yaml]: Done in commit https://github.com/openkruise/kruise/pull/468/commits/aa5bc2c7c7d02e38c348cbbb7134aa9d8b76c80c

File: pkg/controller/advancedcronjob/advancedcronjob_controller_test.go (hunk 1 above)
Comment (rishi-anand) [reply pkg/controller/advancedcronjob/advancedcronjob_controller_test.go]: Done in commit https://github.com/openkruise/kruise/pull/468/commits/aa5bc2c7c7d02e38c348cbbb7134aa9d8b76c80c

File: pkg/webhook/advancedcronjob/mutating/advancedcronjob_create_update_handler.go (hunk 2 above)
Comment (rishi-anand) [reply pkg/webhook/advancedcronjob/mutating/advancedcronjob_create_update_handler.go]: Done in commit https://github.com/openkruise/kruise/pull/468/commits/aa5bc2c7c7d02e38c348cbbb7134aa9d8b76c80c

File: pkg/webhook/advancedcronjob/mutating/webhooks.go (hunk 3 above)
Comment (rishi-anand) [reply pkg/webhook/advancedcronjob/mutating/webhooks.go]: Done in commit https://github.com/openkruise/kruise/pull/468/commits/aa5bc2c7c7d02e38c348cbbb7134aa9d8b76c80c

File: pkg/webhook/advancedcronjob/validating/advancedcronjob_create_update_handler.go (hunk 4 above)
Comment (rishi-anand) [reply pkg/webhook/advancedcronjob/validating/advancedcronjob_create_update_handler.go]: Done in commit https://github.com/openkruise/kruise/pull/468/commits/aa5bc2c7c7d02e38c348cbbb7134aa9d8b76c80c

File: pkg/webhook/advancedcronjob/validating/webhooks.go (hunk 5 above)
Comment (rishi-anand) [reply pkg/webhook/advancedcronjob/validating/webhooks.go]: Done in commit https://github.com/openkruise/kruise/pull/468/commits/aa5bc2c7c7d02e38c348cbbb7134aa9d8b76c80c

File: apis/apps/v1alpha1/advancedcronjob_types.go [hunk 10]
```diff
@@ -43,7 +48,7 @@ type AdvancedCronJobSpec struct {
 
 	// Paused will pause the cron job.
 	// +optional
-	Paused bool `json:"paused,omitempty" protobuf:"bytes,4,opt,name=paused"`
+	Paused *bool `json:"paused,omitempty" protobuf:"bytes,4,opt,name=paused"`
```
Comment (rishi-anand) [inline apis/apps/v1alpha1/advancedcronjob_types.go:51]: Made it to the optional field and the default is `false`.

File: apis/apps/v1alpha1/advancedcronjob_types.go [hunk 11]
```diff
@@ -66,11 +71,11 @@ type AdvancedCronJobSpec struct {
 type CronJobTemplate struct {
 	// Specifies the job that will be created when executing a CronJob.
 	// +optional
-	JobTemplate *batchv1beta1.JobTemplateSpec `json:"jobTemplate" protobuf:"bytes,1,opt,name=jobTemplate"`
+	JobTemplate *batchv1beta1.JobTemplateSpec `json:"jobTemplate,omitempty" protobuf:"bytes,1,opt,name=jobTemplate"`
```
Comment (rishi-anand) [inline apis/apps/v1alpha1/advancedcronjob_types.go:74]: Added `omitempty` for `JobTemplate` and `BroadcastJobTemplate` as it is an optional field.

**Comment by rishi-anand** [comment]
> I don't think you need to add req.NamespacedName in all logs. If the reconciler returns an error, the controller runtime should print out the failed request.

If there are multiple instances of CRD then from the log, we should be able to figure out which log is for which CRD instance.

File: config/crd/bases/apps.kruise.io_advancedcronjobs.yaml [hunk 12]
```diff
@@ -181,6 +181,7 @@ spec:
                   x-kubernetes-preserve-unknown-fields: true
               type: object
           required:
+          - concurrencyPolicy
```
Comment (Fei-Guo) [inline config/crd/bases/apps.kruise.io_advancedcronjobs.yaml:184]: Do you need to remove this as well?

File: config/rbac/role.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fei-Guo) [reply config/rbac/role.yaml:214]: Correct me if I am wrong. 

I don't find RBAC for advancedcronjobs in config/rbac/role.yaml after your change. Is this expected?

[the remark this answers is not in the document]
Comment (rishi-anand) [reply config/rbac/role.yaml:214]: rbac for advancedcronjob is there from 1st PR itself. Here it is https://github.com/openkruise/kruise/blob/e25a1f81b3d201fba1132fe7954ad2dc64f574db/config/rbac/role.yaml#L71-L90

File: config/crd/bases/apps.kruise.io_advancedcronjobs.yaml (hunk 12 above)
Comment (rishi-anand) [reply config/crd/bases/apps.kruise.io_advancedcronjobs.yaml]: Generated manifest again in commit https://github.com/openkruise/kruise/pull/468/commits/038cb120e2e87f5fed7c82e5b8e266d66a02bb58

**Comment by rishi-anand** [comment]
@Fei-Guo, rbac for advancedcronjob is there from 1st PR itself. Here it is https://github.com/openkruise/kruise/blob/e25a1f81b3d201fba1132fe7954ad2dc64f574db/config/rbac/role.yaml#L71-L90

**Comment by Fei-Guo** [comment]
> @Fei-Guo, rbac for advancedcronjob is there from 1st PR itself. Here it is
> 
> https://github.com/openkruise/kruise/blob/e25a1f81b3d201fba1132fe7954ad2dc64f574db/config/rbac/role.yaml#L71-L90

Got you! Sorry, I overlook that.

**Comment by Fei-Guo** [comment]
Please fix the ci failure.

**Comment by rishi-anand** [comment]
> Please fix the ci failure.

Done @Fei-Guo

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**